File: leetcode/198_House_Robber.py
import unittest
from typing import List

class Solution:
    def rob(self, nums: List[int]) -> int:
        dp1, dp2 = 0, 0
        for num in nums:
            dp1, dp2 = dp2 + num, dp1 if dp1 > dp2 else dp2
        return max(dp1, dp2)

class RobTest(unittest.TestCase):
    def test_1(self) -> None:
        self.assertEqual(Solution().rob([1,2,3,1]), 4) 

    def test_2(self) -> None:
        self.assertEqual(Solution().rob([2,7,9,3,1]), 12)

File: leetcode/198_House_Robber.rs
struct Solution {}

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let (mut prev_with_return, mut prev_without_return) = (0, 0);
        for num in nums {
            let with_return = prev_with_return.max(prev_without_return + num);
            let without_return = prev_with_return.max(prev_without_return);
            prev_with_return = with_return;
            prev_without_return = without_return;
        }
        prev_with_return.max(prev_without_return)
    }
}

#[test]
fn test_1() {
    assert_eq!(4, Solution::rob(vec![1,2,3,1]))
}

#[test]
fn test_2() {
    assert_eq!(12, Solution::rob(vec![2,7,9,3,1]))
}
